### PR TITLE
kubernetes_cluster: Add support for private_dns_zone_id

### LINKED ISSUE #10193: Support for Custom Private DNS Zone for AKS
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

BYO Private Zone DNS is now supported: https://docs.microsoft.com/en-us/azure/aks/private-clusters#configure-private-dns-zone

### New or Affected Resource(s)

* azurerm_kubernetes_cluster

### Potential Terraform Configuration

```hcl
resource "azurerm_private_dns_zone" "my_private_dns_zone" {
  name                = "<unique-id-in-east-us>.privatelink.eastus.azmk8s.io"
  resource_group_name = azurerm_resource_group.my_rg.name
}

resource "azurerm_kubernetes_cluster" "aks_private_cluster" {
  name                    = "aks-private-cluster-test"
  location                = "eastus"
  resource_group_name     = azurerm_resource_group.my_rg.name
  dns_prefix              = "..."
  kubernetes_version      = "1.19.0"
  private_cluster_enabled = true
  private_dns_zone_id     = azurerm_private_dns_zone.my_private_dns_zone.id

  lifecycle {
    ignore_changes = [
      # Ignore nodes_count on node pool where autoscaling is on
      default_node_pool[0].node_count,
    ]
  }

  default_node_pool {
    name                = "nodepool1"
    enable_auto_scaling = true
    node_count          = 2
    max_count           = 5
    min_count           = 1
    vm_size             = "Standard_DS2_v2"
    vnet_subnet_id      = azurerm_subnet.my_subnet.id
  }

  identity {
    type = "SystemAssigned"
  }

  linux_profile {
    ...
  }

  addon_profile {
    ...
  }

  network_profile {
    network_plugin     = "azure"
    dns_service_ip     = "192.168.1.10"
    docker_bridge_cidr = "172.17.0.1/16"
    service_cidr       = "192.168.1.0/24"
  }
}
```

### PR DESCRIPTION
**Pull request by favoretti**
Fixes #10193
Blocked on #7979 #8737

### CODE REVIEW

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_cluster_resource.go:443]: could we add validation that this is a Private DNS Zone ID here too?

```suggestion
				ForceNew: true,
				ValidateFunc: privateDnsValidate.PrivateDnsZoneID,
```

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_cluster_resource.go:440]: could we [add this new field to the documentation](https://github.com/terraform-providers/terraform-provider-azurerm/blame/master/website/docs/r/kubernetes_cluster.html.markdown#L101)?

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_cluster_resource.go:804]: an `identity` block is _technically_ optional at this point (although it's required in future) where you can still have a cluster using a Service Principal - so is it worth inverting this:

```suggestion
		if parameters.Identity == nil || parameters.Identity.Type != containerservice.ResourceIdentityTypeUserAssigned {
```

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_cluster_resource.go:805]:
```suggestion
			return fmt.Errorf("a user assigned identity must be used when using a private dns zone")
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @favoretti 

Thanks for this PR :)

I've taken a look through and left a few minor comments inline but this otherwise LGTM - I've merged #8737, so if we can rebase this then the tests should be good to run 👍 

Thanks!

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go
Comment (favoretti) [reply azurerm/internal/services/containers/kubernetes_cluster_resource.go:440]: Yeah, certainly, I wasn't done with this one yet, since I didn't know how long UserAssigned identities will take to merge. I'll finalize this tonight or tomorrow.

**Review by tombuildsstuff — APPROVED** [review]
LGTM 👍

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2021-01-19 at 11 50 51

File: azurerm/internal/services/containers/tests/kubernetes_cluster_network_resource_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/tests/kubernetes_cluster_network_resource_test.go:1244]: looks like this isn't valid:

```
------- Stdout: -------
=== RUN   TestAccAzureRMKubernetesCluster_privateClusterOnWithPrivateDNSZone
=== PAUSE TestAccAzureRMKubernetesCluster_privateClusterOnWithPrivateDNSZone
=== CONT  TestAccAzureRMKubernetesCluster_privateClusterOnWithPrivateDNSZone
testing.go:684: Step 0 error: config is invalid: Unsupported argument: An argument named "identity_ids" is not expected here.
--- FAIL: TestAccAzureRMKubernetesCluster_privateClusterOnWithPrivateDNSZone (0.27s)
FAIL
```

as such I think this wants to become:

```suggestion
    user_assigned_identity_id = azurerm_user_assigned_identity.test.id
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @favoretti 

I've run the tests for this and it appears the configuration is slightly out - but I think if we can update the field name then we should otherwise be good here 👍 

Thanks!

**Comment by favoretti** [comment]
Hey, don't merge this one yet pls. I hear BYO DNS needs some potentially dirty magic, wanna test a hypothesis tomorrow first.

**Comment by favoretti** [comment]
@tombuildsstuff This _should_ work, although I haven't tested all the options.

**Comment by dansanabria** [comment]
We need this! initial testing looks promising

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_cluster_resource.go:826]: this likely only applies when not using "System" - could we add an acctest for this?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @favoretti 

Thanks for pushing those changes - I've taken a look through and this is looking good, if we can remove the `None` option and add an acceptance test covering `System` (which I think'll require adjusting the validation logic in the Create) - this should otherwise be good to go 👍 

Thanks!

**Comment by favoretti** [comment]
```
$ TF_ACC=1 go test -v ./azurerm/internal/services/containers/ -timeout=1000m -run TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem
=== RUN   TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem
=== PAUSE TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem
=== CONT  TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem
--- PASS: TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem (834.53s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers	835.738s
```

**Comment by favoretti** [comment]
```
$ TF_ACC=1 go test -v ./azurerm/internal/services/containers/ -timeout=1000m -run TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZone
=== RUN   TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZone
=== PAUSE TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZone
=== RUN   TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem
=== PAUSE TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem
=== CONT  TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZone
=== CONT  TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem
--- PASS: TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZone (802.86s)
--- PASS: TestAccKubernetesCluster_privateClusterOnWithPrivateDNSZoneSystem (829.75s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers	830.968s
```

**Review by tombuildsstuff — APPROVED** [review]
LGTM 👍

**Comment by tombuildsstuff** [comment]
Running the tests for this, it appears that Private Clusters may need to explicitly set `private_dns_zone_id` to `System`, but that public clusters are unaffected - whilst I've run a subset of the tests, I think we need a full suite run for the Kubernetes package to be sure

**Comment by favoretti** [comment]
Push it to 2.47 otherwise? I can add more acceptance tests as well to see whether not setting it to anything works.

**Comment by tombuildsstuff** [comment]
Yeah, I have a feeling this _only_ affects Private Clusters (I spotted this for `TestAccKubernetesCluster_privateClusterOn`) - I'm running the suite now but I have a feeling this'll need an upgrade note explaining this, so I'm going to push this back to next week's release

**Comment by EppO** [comment]
> Running the tests for this, it appears that Private Clusters may need to explicitly set `private_dns_zone_id` to `System`, but that public clusters are unaffected - whilst I've run a subset of the tests, I think we need a full suite run for the Kubernetes package to be sure

Private AKS clusters **should** be able to use `private_dns_zone_id` with the Resource ID of the private DNS zone per the docs 
> "Custom private dns zone name" should be in this format for azure global cloud: privatelink.<region>.azmk8s.io. You will need the Resource Id of that Private DNS Zone. Additionally, you will need a user assigned identity or service principal with at least the private dns zone contributor role to the custom private dns zone.

So I assume the workflow would be to:

- create a SP or a user assigned identity
- create a private DNS zone
- assign to the SP or user assigned identity a private DNS zone contributor role on the  private DNS zone
- create the AKS cluster with the `private_dns_zone_id` and the SP `client_id/client_secret` or user assigned identity's `user_assigned_identity_id`

**Pull request merged**
